File: combinatorics/resource/ST/project_5/bai_toan_8_9_10/bfs_mutil.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

// Nhap do thi
void input_multigraph(int& n, vector<vector<int>>& adj) {
    int m;
    cout << "Nhap so dinh: ";
    cin >> n;
    cout << "Nhap so canh: ";
    cin >> m;
    adj.resize(n);
    for (int i = 0; i < m; ++i) {
        int u, v;
        cout << "Nhap canh u v (co the trung): ";
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);  // Do thi vo huong, trung canh
    }
}

// BFS
void bfs_multigraph(int start, vector<vector<int>>& adj, int n) {
    vector<bool> visited(n, false);
    queue<int> q;
    visited[start] = true;
    q.push(start);
    cout << "Thu tu duyet: ";
    while (!q.empty()) {
        int u = q.front(); q.pop();
        cout << u << " ";
        for (int v : adj[u]) {
            if (!visited[v]) {
                visited[v] = true;
                q.push(v);
            }
        }
    }
}

int main() {
    int n;
    vector<vector<int>> adj;
    input_multigraph(n, adj);
    int s;
    cout << "Nhap dinh bat dau: ";
    cin >> s;
    bfs_multigraph(s, adj, n);
    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_8_9_10/bfs_mutil.py
from collections import deque

# Nhap da do thi tu ban phim
def input_multigraph():
    n = int(input("Nhap so dinh: "))
    m = int(input("Nhap so canh: "))
    adj = [[] for _ in range(n)]
    for _ in range(m):
        u, v = map(int, input("Nhap canh u v (co the trung): ").split())
        adj[u].append(v)
        adj[v].append(u)  # Do thi vo huong co the trung canh
    return n, adj

# BFS co xu ly trung canh
def bfs_multigraph(start, adj, n):
    visited = [False] * n
    queue = deque([start])
    visited[start] = True
    print("Thu tu duyet:", end=" ")
    while queue:
        u = queue.popleft()
        print(u, end=" ")
        for v in adj[u]:
            if not visited[v]:  # Chi dua vao hang doi neu chua tham
                visited[v] = True
                queue.append(v)

# Main
if __name__ == "__main__":
    n, adj = input_multigraph()
    s = int(input("Nhap dinh bat dau: "))
    bfs_multigraph(s, adj, n)